### ISSUE TITLE
cp: cannot stat 'obj/mbld/mbld': No such file or directory

### ISSUE THREAD

**Issue by Earnestly** [oridb/mc#184]
* https://github.com/oridb/mc/commit/a819f7ccdc3ed6c63f50f23c0b17fa478604d821

Building after a clean clone with the following:
```
% ./configure --prefix=/usr
% make -j4 bootstrap            # This happens without -j4 as well
```

Yields a new error:

```
[...]
+ ar -rcs obj/lib/regex/libregex.a obj/lib/regex/interp.o obj/lib/regex/ranges.o obj/lib/regex/types.o obj/lib/regex/compile.o
+ /home/earnest/build/staging/myrddin-git/src/mc/muse/muse -o obj/lib/regex/libregex.use -p regex obj/lib/regex/interp.use obj/lib/regex/ranges.use obj/lib/regex/types.use obj/lib/regex/compile.use
+ /home/earnest/build/staging/myrddin-git/src/mc/6/6m -O obj -I obj/lib/sys -I obj/lib/std -I obj/lib/bio -I obj/lib/regex -I obj/lib/thread mbld/subtest.myr
+ /home/earnest/build/staging/myrddin-git/src/mc/6/6m -O obj -I obj/lib/sys -I obj/lib/std -I obj/lib/bio -I obj/lib/regex -I obj/lib/thread mbld/test.myr
+ /home/earnest/build/staging/myrddin-git/src/mc/6/6m -O obj -I obj/lib/sys -I obj/lib/std -I obj/lib/bio -I obj/lib/regex -I obj/lib/thread mbld/deps.myr
+ /home/earnest/build/staging/myrddin-git/src/mc/6/6m -O obj -I obj/lib/sys -I obj/lib/std -I obj/lib/bio -I obj/lib/regex -I obj/lib/thread mbld/main.myr
+ ld --gc-sections -o obj/mbld/mbld /home/earnest/build/staging/myrddin-git/src/mc/rt/_myrrt.o obj/mbld/deps.o obj/mbld/main.o obj/mbld/util.o obj/mbld/cpufeatures.o obj/mbld/libs.o obj/mbld/syssel.o obj/mbld/config.o obj/mbld/opts.o obj/mbld/subtest.o obj/mbld/types.o obj/mbld/test.o obj/mbld/install.o obj/mbld/parse.o obj/mbld/build.o -Lobj/lib/thread -lthread -Lobj/lib/bio -lbio -Lobj/lib/regex -lregex -Lobj/lib/std -lstd -Lobj/lib/sys -lsys
ld: cannot find -lthread
ld: cannot find -lstd
ld: cannot find -lsys
+ true
cp obj/mbld/mbld xmbld
cp: cannot stat 'obj/mbld/mbld': No such file or directory
make: *** [Makefile;26: bootstrap] Error 1
```

Interestingly, if I run `make bootstrap` again, it completes without issue.

**Comment by oridb** [comment]
Hm. What platform? (The change you cited only affects plan 9).

If you install, and then run `make genbootstrap`, do you get differences? If you do, can you send a pull request?

**Comment by kirbyfan64** [comment]
> Ah, I thought you may remember me.

Something about this is just so sad...

**Comment by Earnestly** [comment]
I've decided to just include `make genbootstrap` in the usual build as it avoids this issue and any future issues that might originate from improperly generated bootstraps.

**Comment by oridb** [comment]
Keep in mind that genbootstrap assumes you have a working compiler already available. Bootstrap is there only to get you to a working compiler and build system.

At some point, I'm thinking of just having the bootstrap pull down prebuilt/presigned tarballs.

**Comment by Earnestly** [comment]
Oh, that's no good then.  Ideally the bootstrap should not download anything.  It should be possible to download everything necessary for a build and then not require network access beyond that point.

**Comment by oridb** [comment]
The root of the problem is that the libraries are built with mbld; Mbld depends on the libraries. The bootstrap script hard codes a line by line replay of what the mbld  that generated the bootstrap script did.

We need *some* way of getting a working mbld to build things. That mbld can come from anywhere, but it needs to come from somewhere. Options include a `-bootstrap` tarball, be generated via the `mk/bootstrap/*` script, be committed into the repo for each architecture, or be baked into a release tarball that's used to build from git.

I'm not sure what's the least painful.

**Comment by oridb** [comment]
(Also, I definitely remember you, but I don't memorize what platform everyone uses -- and even if I did, people sometimes molt, and come back with a new computer.)

**Comment by Earnestly** [comment]
Heh, fair enough.  Not sure what to suggests here but keeping tarballs in the git repo sounds like a bad call.  The chezscheme guys did this and eventually ended up with a repo over 1G in size.

**Comment by kirbyfan64** [comment]
What about just doing a pre-commit hook that makes sure the bootstrap is up to date?

**Comment by akkartik** [comment]
https://github.com/oridb/mc#build
> If you are building from development then issue `make bootstrap` after the call to `./configure`.

https://myrlang.org
> When running from git, you may need to first run `make bootstrap` on the first installation, or upon significant changes to `mbld`.

Probably a stupid question, but what is the alternative to building from git that doesn't require running `make bootstrap`?

**Comment by oridb** [comment]
Have an old version available.

**Comment by kirbyfan64** [comment]
@akkartik AFAIK the bootstrap scripts were already fixed/regenerated on master; this bug is more about the long-term solving of the issue of the build scripts constantly going out of sync.

**Comment by oridb** [comment]
Correct: `make bootstrap` should work on all platforms except NetBSD at the moment. I haven't had a chance to correct that one yet.

For robustness, I'll probably figure out some sort of continuous integration + commit hook that rewrites the bootstrap soon. The annoyance is that it's going to have to be an external system, because there are a number of platforms that need this. I'll try to get to that in the next week or to.

**Comment by akkartik** [comment]
Ah, that makes sense thanks. I found this issue because I ran into that `cp: cannot stat 'obj/mbld/mbld': No such file or directory` error message. I'm on Mac OS X:

```
$ uname -a
Darwin 4133 18.2.0 Darwin Kernel Version 18.2.0: Mon Nov 12 20:24:46 PST 2018; root:xnu-4903.231.4~2/RELEASE_X86_64 x86_64 i386 MacBookPro14,1 Darwin
```

The steps I ran were:
```
$ git clone https://github.com/oridb/mc
$ cd mc
$ ./configure
$ make bootstrap
$ make bootstrap
```

As you can see, rerunning `make bootstrap` didn't fix the problem like it seems to have done for @Earnestly above. Let me know if you'd like me to create a fresh issue for this.

**Comment by akkartik** [comment]
The same commit hash (261918774d46c50c96f729f4e92b01323c56c398) bootstraps fine on Ubuntu 18.04.1. I hadn't tried that before, thanks.

**Comment by Earnestly** [comment]
The issue has occurred again.  I like @kirbyfan64 suggestion of using commit hooks or something of that nature because it would be a great shame to loose such an effortless bootstrap, one better than any other language and runtime I've had to compile.

As a workaround I've decided to make `genbootstrap` conditional on the presence of `mbld` in the system.  This is only temporary as each package should be self-contained.

**Comment by oridb** [comment]
I think that in the next few days, I'll set up a few machines to do the build and regeneration on.

**Comment by oridb** [comment]
Should be 90% done now -- the bootstrap script to regenerate across all platforms is committed, but it needs a build farm to run the bootstrap. Some of the machines are local -- partly because I haven't yet set them up on a vultr node, but the OSX machine will probably stay where it is because, well, there are no good hosting providers for mac infrastructure.

It may end up in a colo at some point.

**Comment by lojikil** [comment]
I'm seeing a related issue on OS X:

```
+ ld -pagezero_size 0x100000000 -macosx_version_min 10.6 -o obj/mbld/mbld /Users/lojikil/Build/mc/rt/_myrrt.o obj/mbld/deps.o obj/mbld/main.o obj/mbld/util.o obj/mbld/cpufeatures.o obj/mbld/libs.o obj/mbld/syssel.o obj/mbld/config.o obj/mbld/opts.o obj/mbld/subtest.o obj/mbld/types.o obj/mbld/test.o obj/mbld/install.o obj/mbld/parse.o obj/mbld/build.o -Lobj/lib/thread -lthread -Lobj/lib/bio -lbio -Lobj/lib/regex -lregex -Lobj/lib/std -lstd -Lobj/lib/sys -lsys -macosx_version_min 10.6
ld: warning: passed two min versions (10.6.0, 10.6) for platform macOS. Using 10.6.
Undefined symbols for architecture x86_64:
  "_std$intfmt", referenced from:
      .L163 in libstd.a(fmt.o)
      .L164 in libstd.a(fmt.o)
      .L165 in libstd.a(fmt.o)
      .L166 in libstd.a(fmt.o)
      .L167 in libstd.a(fmt.o)
      .L168 in libstd.a(fmt.o)
      .L169 in libstd.a(fmt.o)
      ...
ld: symbol(s) not found for architecture x86_64
+ true
cp obj/mbld/mbld xmbld
cp: obj/mbld/mbld: No such file or directory
make: *** [bootstrap] Error 1
lojikil@malum [~/Build/mc]
% uname -a  
Darwin malum.local 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64 x86_64
```

The build sequence was:

```
git clone ...
./configure --prefix=/usr/local
make bootstrap
```

which then failed with the above

**Comment by Earnestly** [comment]
@oridb Perhaps consider using https://man.sr.ht/builds.sr.ht/ for generating bootstraps on multiple platforms?

**Comment by oridb** [comment]
Yeah, that's a good idea. I'll probably try to get to it next weekend.
